File: algos/pacific-atlantic/pacific-atlantic.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        """
        approach:
        BFS from every cell adjacent to pacific
        BFS from every cell adjacent to atlantic

        can recursively move to a cell with higher or same value
        if any 2 cells from BOTH BFS converge, record convergent cell
        """
        ROWS, COLS = len(heights), len(heights[0])

        pacific, atlantic = set(), set()

        # prevHeight lets us check if we are allowed to visit this cell
        def dfs(r, c, visit, prevHeight):
            if ((r,c) in visit or 
            r < 0 or c < 0 or r == ROWS or c == COLS or
            heights[r][c] < prevHeight):
                return
            visit.add((r,c))
            dfs(r+1,c,visit,heights[r][c])
            dfs(r-1,c,visit,heights[r][c])
            dfs(r,c+1,visit,heights[r][c])
            dfs(r,c-1,visit,heights[r][c])
        
        for c in range(COLS):
            # go through every pos in first row (touching pacific)
            dfs(0, c, pacific, heights[0][c])
            # go through every pos in last row (touching atlantic)
            dfs(ROWS-1, c, atlantic, heights[ROWS-1][c])
        
        for r in range(ROWS):
            # go through every pos in first col (touching pacific)
            dfs(r, 0, pacific, heights[r][0])
            # go through every pos in last col (touching atlantic)
            dfs(r, COLS-1, atlantic, heights[r][COLS-1])
        
        res = []
        for r in range(ROWS):
            for c in range(COLS):
                if (r,c) in pacific and (r,c) in atlantic:
                    res.append([r, c])
        return res

File: algos/pacific-atlantic/pacific-atlantic.js
class Solution {
  pacificAtlantic(heights) {
    const ROWS = heights.length;
    const COLS = heights[0].length;
    const pacific = new Set();
    const atlantic = new Set();

    for (let i = 0; i < ROWS; i++) {
      this.dfs(i, 0, pacific, heights[i][0], ROWS, COLS, heights);
      this.dfs(
        i,
        COLS - 1,
        atlantic,
        heights[i][COLS - 1],
        ROWS,
        COLS,
        heights
      );
    }

    for (let j = 0; j < COLS; j++) {
      this.dfs(0, j, pacific, heights[0][j], ROWS, COLS, heights);
      this.dfs(
        ROWS - 1,
        j,
        atlantic,
        heights[ROWS - 1][j],
        ROWS,
        COLS,
        heights
      );
    }

    const res = [];
    for (let i = 0; i < ROWS; i++) {
      for (let j = 0; j < COLS; j++) {
        const coord = i * COLS + j; // generate unique val
        if (pacific.has(coord) && atlantic.has(coord)) {
          res.push([i, j]);
        }
      }
    }
    return res;
  }

  dfs(r, c, visited, prevHeight, ROWS, COLS, heights) {
    const coord = r * COLS + c; // generate unique val
    if (
      r < 0 ||
      c < 0 ||
      r == ROWS ||
      c == COLS ||
      heights[r][c] < prevHeight ||
      visited.has(coord)
    )
      return;

    visited.add(coord);
    this.dfs(r + 1, c, visited, heights[r][c], ROWS, COLS, heights);
    this.dfs(r - 1, c, visited, heights[r][c], ROWS, COLS, heights);
    this.dfs(r, c + 1, visited, heights[r][c], ROWS, COLS, heights);
    this.dfs(r, c - 1, visited, heights[r][c], ROWS, COLS, heights);
  }
}
